### PR TITLE
Ripe cleanup code and restructure the logic

### PR DESCRIPTION
**Pull request by Teebor-Choka**
Complete overhaul of RIPE expert to remove all errors/problems in logs and code.

### CODE REVIEW

**Comment by wagner-certat** [comment]
Which problems did you have with the RIPE API? Weren't the responses sane?

**Comment by Teebor-Choka** [comment]
@wagner-certat API is good, the handling of incoming data was generating errors in logs.

The culprit was (primarily) negative case handling: e.g. in geolocation query the code would attempt to access a list element, which would fail with `IndexError`, because some records did not have any geolocation data within the reply. The generated exception was not `except`ed in the RIPE expert, so the record would eventually end up in a `.dump` file, instead of being passed along the pipeline.

While fixing smaller bugs like this I decided to refactor + streamline the code and remove code duplication.

**Comment by Teebor-Choka** [comment]
A particular nuisance was the presence of `"?"` inside a RIPE reply indicating that RIPE could not establish information regarding the country geolocation (latitude and longitude would point to the international waters, Antarctica, etc...). `RIPEExpert` did not handle such cases at all before, now it properly handles the case by not filling any data into the `*.geolocation.cc` field.

File: intelmq/bots/experts/ripe/expert.py [hunk 1]
```diff
@@ -15,177 +17,144 @@
...
+                        should_overwrite = self.__mode == 'replace'
+
+                        for local_key, ripe_key in self.GEOLOCATION_REPLY_TO_INTERNAL:
+                            if ripe_key in info:
+                                event.add(target + "geolocation." + local_key, info[ripe_key], overwrite=should_overwrite)
+
+                event.add(abuse_key, ','.join(abuse), overwrite=True)
+
+    @contextmanager
+    def event_context(self):
```
Comment (Teebor-Choka) [inline intelmq/bots/experts/ripe/expert.py]: This function should actually be moved into the `parent` class, where it should also be used.

**Comment by wagner-certat** [comment]
> The culprit was (primarily) negative case handling: e.g. in geolocation query the code would attempt to access a list element, which would fail with `IndexError`, because some records did not have any geolocation data within the reply. The generated exception was not `except`ed in the RIPE expert, so the record would eventually end up in a `.dump` file, instead of being passed along the pipeline.

Do you still know a resource which triggers this response, so we can use it in the tests?

> 
> While fixing smaller bugs like this I decided to refactor + streamline the code and remove code duplication.

Yeah, thats great :)

File: intelmq/bots/experts/ripe/expert.py [hunk 2]
```diff
@@ -15,177 +17,144 @@
...
-            return {}
-        response = self.http_session.get(self.URL_STAT_GEOLOCATION.format(ip),
-                                         timeout=self.http_timeout_sec)
-        if response.status_code != 200:
-            raise ValueError(STATUS_CODE_ERROR % response.status_code)
+                return json.loads(cached_value)
+        else:
+            response = self.http_session.get(self.QUERY[type].format(resource), data="", timeout=self.http_timeout_sec)
+            if response.status_code != 200:
+                if type != 'db_asn' or response.status_code != 404:     # 404 => contact not found
```
[the remark this answers is not in the document]
Comment (Teebor-Choka) [reply intelmq/bots/experts/ripe/expert.py]: On the contrary, for `db_asn` the `404` code represents "contact not found", not "URL not found". Therefore it is discarded. This behavior is taken from the original `RIPEExpert`, the code was just refactored. If I remember correctly, there even was a specific test for this case, so it appears to be a specialty of `db_asn`. Dunno if that applies to all `db_*`.

File: intelmq/bots/experts/ripe/expert.py [hunk 3]
```diff
@@ -15,177 +17,144 @@
...
-        cache_value = self.cache.get('stat:%s' % resource)
-        if cache_value and cache_value != CACHE_NO_VALUE:
-            return json.loads(cache_value)
-        elif cache_value == CACHE_NO_VALUE:
-            return []
-        response = self.http_session.get(self.URL_STAT_CONTACT.format(resource), data="",
-                                         timeout=self.http_timeout_sec)
-        if response.status_code != 200:
-            raise ValueError(STATUS_CODE_ERROR % response.status_code)
+            self.__cache = Cache(cache_host, cache_port, cache_db, cache_ttl,
```
[the remark this answers is not in the document]
Comment (Teebor-Choka) [reply intelmq/bots/experts/ripe/expert.py]: To enforce cache usage through this expert's API (currently none) and prohibit external elements to tamper with the cache.

**Comment by Teebor-Choka** [comment]
> Do you still know a resource which triggers this response, so we can use it in the tests?

`IndexError` on invalid array access [case](https://stat.ripe.net/data/maxmind-geo-lite/data.json?resource=228.66.141.189)
`"?"` [case](https://stat.ripe.net/data/maxmind-geo-lite/data.json?resource=35.197.157.0)

I no longer remember other fixes I did there. :1234:

File: intelmq/bots/experts/ripe/expert.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wagner-certat) [reply intelmq/bots/experts/ripe/expert.py]: No, previously, for a 404 no exception has been raised for DB AS queries, but for all other status codes. Now it is the other way round. Interestingly, the `test_ripe_db_as_404` test succeeds. And further only a small portion of the RIPE tests is run on travis.... wtf
The coverage here is only 18% for the expert: https://travis-ci.org/certtools/intelmq/jobs/502550013#L1037
Running the tests locally gives 90%
In need to debug the tests problem.

This 404 case has been introduced in ca6e6756b1c8341de6ee2a8f3113f94866e85ed6 btw

[the remark this answers is not in the document]
Comment (wagner-certat) [reply intelmq/bots/experts/ripe/expert.py]: Could find the culprit why the ripe tests have never been run: a missing `__init__.py` (though these files are not necessary on python 3 anymore). Fixed in maintenance and develop, see 9b8cf71203901ca87e07fc1bdbc34d3a8ec783cb

[the remark this answers is not in the document]
Comment (Teebor-Choka) [reply intelmq/bots/experts/ripe/expert.py]: > No, previously, for a 404 no exception has been raised for DB AS queries, but for all other status codes. Now it is the other way round. Interestingly, the `test_ripe_db_as_404` test succeeds.

I'm getting lost in your thoughts, but regardless, is the refactored code valid or not? The tests should provide an unambiguous answer to that.

[the remark this answers is not in the document]
Comment (wagner-certat) [reply intelmq/bots/experts/ripe/expert.py]: > I'm getting lost in your thoughts, but regardless, is the refactored code valid or not? The tests should provide an unambiguous answer to that.

I worked on fixing the bugs/test fails which I revealed after the discovery the the missing files causes some test cases not to execute. I will work on you PR tomorrow and I guess I'll merge it and eventually adapt some things myself (the 404 thing and the tests)

**Comment by wagner-certat** [comment]
I cherry-picked the commit into develop (there are too many unrelated merges in your branch and some duplicated earlier commits). My changes are in b38a68d628c3af63f78b66f40a107a73956b0063

**Pull request closed without merging**